immutable.d.ts

```diff
@@ -21,6 +21,20 @@ interface KeyValuePair<K, V> extends Array<K | V> {
   readonly 1: V;
 }
 
+type Prev = [never, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, ...0[]]
+
+type Join<K, P> = K extends string | number ?
+  P extends string | number ?
+    `${K}${"" extends P ? "" : "."}${P}`
+    : never : never;
+
+// Prev, Join, and Paths are based on discussion in SO: https://stackoverflow.com/questions/58434389/typescript-deep-keyof-of-a-nested-object
+type Paths<T, D extends number = 10> = [D] extends [never] ? never : T extends object ?
```

Comment:
I read the discussion and I am a bit worried about this part:

`Be warned: Paths and Leaves are inherently recursive in a way that can be very taxing on the compiler. `

Reply:
Welp, I wanted to create a file to test some of these large scenarios, but I can't get my code to work now as the screenshot showed. I'm quite perplexed.

Reply:
Not really sure how to test compiler speeds, but at least from my initial testing, the compiler seems to pick it up fine even when deeply nested.

Reply:
yes, we'll need a testing for that - I'll see what I can do in this regard.